File: leetcode/lc-us/cpp/2360.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int longestCycle(vector<int> &edges) {
        int n = (int) edges.size();
        vector<int> time(n);
        int res = -1;
        for (int i = 0, clock = 1; i < edges.size(); i++) {
            if (time[i]) continue;
            for (int x = i, start = clock; ~x; x = edges[x]) {
                if (time[x]) {
                    if (time[x] >= start) {
                        res = max(res, clock - time[x]);
                    }
                    break;
                }
                time[x] = clock++;
            }
        }
        return res;
    }
};

File: leetcode/lc-us/rust/2360.rs
use std::cmp::max;

impl Solution {
    pub fn longest_cycle(edges: Vec<i32>) -> i32 {
        let n = edges.len();
        let mut res: i32 = -1;
        let mut time = vec![0; n];
        let mut start = 1;
        for i in 0..n {
            if time[i] == 0 {
                let mut clock = start;
                let mut x = i as i32;
                while x != -1 {
                    if time[x as usize] != 0 {
                        if time[x as usize] >= start {
                            res = max(res, (clock - time[x as usize]) as i32);
                        }
                        break;
                    }
                    time[x as usize] = clock;
                    clock += 1;
                    x = edges[x as usize];
                }
                start = clock;
            }
        }
        res
    }
}